crates/solvers/src/boundary/naive.rs

```diff
@@ -0,0 +1,235 @@
+) -> Option<solution::Solution> {
+    let pool = match &liquidity.state {
+        liquidity::State::ConstantProduct(pool) => pool,
+        _ => return None,
+    };
+
+    let slippage = Slippage::new(pool.tokens());
+    let capture = Arc::new(Mutex::new(Capture::default()));
+
+    let boundary_orders = orders
+        .iter()
+        .map(|order| LimitOrder {
+            id: match order.class {
+                order::Class::Market => LimitOrderId::Market(OrderUid(order.uid.0)),
+                order::Class::Limit => LimitOrderId::Limit(OrderUid(order.uid.0)),
+                order::Class::Liquidity => {
+                    LimitOrderId::Liquidity(LiquidityOrderId::Protocol(OrderUid(order.uid.0)))
+                }
+            },
+            sell_token: order.sell.token.0,
+            buy_token: order.buy.token.0,
+            sell_amount: order.sell.amount,
+            buy_amount: order.buy.amount,
+            kind: match order.side {
+                order::Side::Buy => OrderKind::Buy,
+                order::Side::Sell => OrderKind::Sell,
+            },
+            partially_fillable: order.partially_fillable,
+            solver_fee: order.fee().amount,
+            is_mature: true,
+            settlement_handling: Arc::new(OrderHandler {
+                order: order.uid,
+                capture: capture.clone(),
+            }),
+            exchange: Exchange::GnosisProtocol,
+            reward: 0.,
+        })
+        .collect_vec();
+    let boundary_pool = ConstantProductOrder::for_pool(
+        to_boundary_pool(liquidity.address, pool)?,
+        Arc::new(PoolHandler {
+            capture: capture.clone(),
+        }),
+    );
+
+    let boundary_solution =
+        multi_order_solver::solve(&slippage.context(), boundary_orders, &boundary_pool)?;
+
+    let mut capture = capture.lock().unwrap();
+    Some(solution::Solution {
+        prices: solution::ClearingPrices::new(
+            boundary_solution
+                .clearing_prices()
+                .iter()
+                .map(|(token, price)| (eth::TokenAddress(*token), *price)),
+        ),
+        trades: orders
+            .iter()
+            .copied()
+            .filter(|order| capture.orders.contains(&order.uid))
+            .map(|order| solution::Trade::Fulfillment(solution::Fulfillment::fill(order.clone())))
+            .collect(),
+        interactions: capture
+            .swap
+            .take()
+            .into_iter()
+            .map(|(input, output)| {
+                solution::Interaction::Liquidity(solution::LiquidityInteraction {
+                    liquidity: liquidity.clone(),
+                    input,
+                    output,
+                    internalize: false,
+                })
+            })
+            .collect(),
+    })
+}
+
+// Beyond this point is... well... nameless and boundless chaos. The
+// unfathomable horrors that follow are not for the faint of heart!
```

Comment:
Sounds like the implementation would be a mess. Do you think it might finally be time to drop the naive solver? AFAIK we were considering that for quite some time now.

Reply:
I was planning on porting it to the `solvers` crate (i.e. moving it from the `boundary` module into the `domain`). I just wanted to setup tests before doing that to make sure there are no bugs/regressions that get introduced. Its actually a pretty small bit of code, and most of the nasty bugs are related to the fact that the code acts on both `LimitOrder` (i.e. the liquidity abstraction) and `Settlement` (i.e. actual orders). By moving it here, we would have a consistent view of liquidity and orders and, in theory, avoid all of the issues we've been seeing with it lately.